scripts/linters/pre_commit_linter.py

```diff
@@ -328,32 +350,99 @@ def _get_file_extensions(file_extensions_to_lint):
     return all_file_extensions_type
 
 
-def _get_all_filepaths(input_path, input_filenames):
+def _get_filepaths_from_path(input_path):
+    """Get paths to all lintable files recursively under a path.
+
+    This function applies some ignore rules (from .eslintignore) but not
+    all.
+
+    Args:
+        input_path: str. Path to look for files under.
+
+    Returns:
+        list. Paths to lintable files.
+    """
+    input_path = os.path.join(os.getcwd(), input_path)
+    if not os.path.exists(input_path):
+        python_utils.PRINT(
+            'Could not locate file or directory %s. Exiting.' % input_path)
+        python_utils.PRINT('----------------------------------------')
+        sys.exit(1)
+    if os.path.isfile(input_path):
+        return [input_path]
+    else:
+        eslintignore_path = os.path.join(os.getcwd(), '.eslintignore')
+        excluded_glob_patterns = FILE_CACHE.readlines(eslintignore_path)
+        return _get_all_files_in_directory(
+            input_path, excluded_glob_patterns)
+
+
+def _get_filepaths_from_non_other_shard(shard):
+    """Get paths to lintable files in a shard besides the other shard.
+
+    This function applies some ignore rules (from .eslintignore) but not
+    all.
+
+    Args:
+        shard: str. Shard name.
+
+    Returns:
+        list(str). Paths to lintable files.
+    """
+    filepaths = []
+    assert shard != OTHER_SHARD_NAME
+    for filepath in SHARDS[shard]:
+        filepaths.extend(_get_filepaths_from_path(filepath))
+    if len(filepaths) != len(set(filepaths)):
+        # Shards are invalid because of a duplicate file.
+        for filepath in filepaths:
+            if filepaths.count(filepath) > 1:
+                raise RuntimeError(
+                    '%s in multiple shards.' % filepath)
+        # We exempt this line from test coverage because it is
+        # un-testable. It should never be reached, but we raise an
+        # assertion error to catch coding errors above.
```

Comment:
We can swap/mock the values for the shard files and test this error, right?

Reply:
I don't think so. Here's the relevant code:

```python
    if len(filepaths) != len(set(filepaths)):
        # Shards are invalid because of a duplicate file.
        for filepath in filepaths:
            if filepaths.count(filepath) > 1:
                raise RuntimeError(
                    '%s in multiple shards.' % filepath)
        # ...
        raise AssertionError(  # pragma: no cover
            'There is a file duplicated across shards. '
            'We should have been able to find it but failed.')
```

We only reach the assertion error if both of the following are true:

1. `len(filepaths) != len(set(filepaths))`: This can only happen if there is a duplicate element in `filepaths`
2. For every `filepath` in `filepaths`, `filepaths.count(filepath) == 1`: This can only happen when there are no duplicates in `filepaths`

Therefore, we should never be able to reach the assertion error. I only included it as a defensive measure to declare my assumption that if `len(filepaths) != len(set(filepaths))`, we raise a `RuntimeError`